# Loading and inspecting data

The first step in the reconstruction stage is to load the aligned data from file.
For this purpose `mumott` provides the `DataContainer` class.
Objects of this class are initialized by reading data from a file, and then hold all relevant data in one place.
This tutorial illustrates this functionality and demonstrates how the data can be queried and transformed after loading.

The simulated data used in this tutorial can be obtained using, e.g., ``wget`` with
```shell
wget https://zenodo.org/record/7326784/files/saxstt_dataset_M.h5
```

We start by importing the `DataContainer` and `TransformParameters` classes.
The latter holds the information pertaining to the coordinate system in which the data was measured.
It can be used to transform the data into a different coordinate system for the reconstruction.

In [1]:
import numpy as np
from mumott.data_handling import DataContainer, TransformParameters

## The `DataContainer` object

We can load the data file into a data container as follows.

In [2]:
dc = DataContainer(data_path='.', data_filename='saxstt_dataset_M.h5')

There are now various options to query the data in the container.
We can for example print the data container, which produces a string represetantion that displays key properties.

In [3]:
print(dc)

                              DataContainer                               
--------------------------------------------------------------------------
Number of projections      : 417
Volume shape               : [50 50 50]
Corrected for transmission : False
Angles in radians          : True
Transformation applied     : False
Reconstruction parameters generated : False


In a jupyter hub environment one can also generate a more nicely formatted version using the display command.
If one "calls" an object.

<div class="alert alert-block alert-info">
<b>Tip:</b> If an object is "invoked" directly and on the last line of a cell, the <code>display</code> command is implied, i.e., <code>display(obj)</code> and <code>obj</code> lead to the same output.
Below we use this short cut to display objects.
</div>

In [4]:
display(dc)

In this example, the data container contains 417 projections.
In its current state, the data has not yet been corrected for transmission and we have not yet applied a transformation.
We can also see that the reconstruction parameters have not yet been generated.
This is triggered internally, once the data is to be used further.
The transmission correction and the transformation must be applied before this step, as will be described below.

## The stack
The data in the container is stored in the form of a stack, which comprises individual frames, each representing a different projection.
In our example the stack contains 417 projections, which we can confirm by printing the length of the stack.

In [5]:
len(dc.stack)

417

<div class="alert alert-block alert-warning">
<b>Note:</b> <code>dc.stack</code> provides a <i>view</i> of the stack (not a copy), to prevent direct (uncontrolled) manipulation of the data.
</div>

We can obtain a convenient overview of the stack in the same fashion as for the data container by calling the object (or printing it when working with a standard interpreter).

In [6]:
dc.stack

InputStack(_data_start=array([      0,   20000,   40000,   60000,   80000,  100000,  120000,
        140000,  160000,  180000,  200000,  220000,  240000,  260000,
        280000,  300000,  320000,  340000,  360000,  380000,  400000,
        420000,  440000,  460000,  480000,  500000,  520000,  540000,
        560000,  580000,  600000,  620000,  640000,  660000,  680000,
        700000,  720000,  740000,  760000,  780000,  800000,  820000,
        840000,  860000,  880000,  900000,  920000,  940000,  960000,
        980000, 1000000, 1020000, 1040000, 1060000, 1080000, 1100000,
       1120000, 1140000, 1160000, 1180000, 1200000, 1220000, 1240000,
       1260000, 1280000, 1300000, 1320000, 1340000, 1360000, 1380000,
       1400000, 1420000, 1440000, 1460000, 1480000, 1500000, 1520000,
       1540000, 1560000, 1580000, 1600000, 1620000, 1640000, 1660000,
       1680000, 1700000, 1720000, 1740000, 1760000, 1780000, 1800000,
       1820000, 1840000, 1860000, 1880000, 1900000, 1920000, 1940000,
       1960000, 1980000, 2000000, 2020000, 2040000, 2060000, 2080000,
       2100000, 2120000, 2140000, 2160000, 2180000, 2200000, 2220000,
       2240000, 2260000, 2280000, 2300000, 2320000, 2340000, 2360000,
       2380000, 2400000, 2420000, 2440000, 2460000, 2480000, 2500000,
       2520000, 2540000, 2560000, 2580000, 2600000, 2620000, 2640000,
       2660000, 2680000, 2700000, 2720000, 2740000, 2760000, 2780000,
       2800000, 2820000, 2840000, 2860000, 2880000, 2900000, 2920000,
       2940000, 2960000, 2980000, 3000000, 3020000, 3040000, 3060000,
       3080000, 3100000, 3120000, 3140000, 3160000, 3180000, 3200000,
       3220000, 3240000, 3260000, 3280000, 3300000, 3320000, 3340000,
       3360000, 3380000, 3400000, 3420000, 3440000, 3460000, 3480000,
       3500000, 3520000, 3540000, 3560000, 3580000, 3600000, 3620000,
       3640000, 3660000, 3680000, 3700000, 3720000, 3740000, 3760000,
       3780000, 3800000, 3820000, 3840000, 3860000, 3880000, 3900000,
       3920000, 3940000, 3960000, 3980000, 4000000, 4020000, 4040000,
       4060000, 4080000, 4100000, 4120000, 4140000, 4160000, 4180000,
       4200000, 4220000, 4240000, 4260000, 4280000, 4300000, 4320000,
       4340000, 4360000, 4380000, 4400000, 4420000, 4440000, 4460000,
       4480000, 4500000, 4520000, 4540000, 4560000, 4580000, 4600000,
       4620000, 4640000, 4660000, 4680000, 4700000, 4720000, 4740000,
       4760000, 4780000, 4800000, 4820000, 4840000, 4860000, 4880000,
       4900000, 4920000, 4940000, 4960000, 4980000, 5000000, 5020000,
       5040000, 5060000, 5080000, 5100000, 5120000, 5140000, 5160000,
       5180000, 5200000, 5220000, 5240000, 5260000, 5280000, 5300000,
       5320000, 5340000, 5360000, 5380000, 5400000, 5420000, 5440000,
       5460000, 5480000, 5500000, 5520000, 5540000, 5560000, 5580000,
       5600000, 5620000, 5640000, 5660000, 5680000, 5700000, 5720000,
       5740000, 5760000, 5780000, 5800000, 5820000, 5840000, 5860000,
       5880000, 5900000, 5920000, 5940000, 5960000, 5980000, 6000000,
       6020000, 6040000, 6060000, 6080000, 6100000, 6120000, 6140000,
       6160000, 6180000, 6200000, 6220000, 6240000, 6260000, 6280000,
       6300000, 6320000, 6340000, 6360000, 6380000, 6400000, 6420000,
       6440000, 6460000, 6480000, 6500000, 6520000, 6540000, 6560000,
       6580000, 6600000, 6620000, 6640000, 6660000, 6680000, 6700000,
       6720000, 6740000, 6760000, 6780000, 6800000, 6820000, 6840000,
       6860000, 6880000, 6900000, 6920000, 6940000, 6960000, 6980000,
       7000000, 7020000, 7040000, 7060000, 7080000, 7100000, 7120000,
       7140000, 7160000, 7180000, 7200000, 7220000, 7240000, 7260000,
       7280000, 7300000, 7320000, 7340000, 7360000, 7380000, 7400000,
       7420000, 7440000, 7460000, 7480000, 7500000, 7520000, 7540000,
       7560000, 7580000, 7600000, 7620000, 7640000, 7660000, 7680000,
       7700000, 7720000, 7740000, 7760000, 7780000, 7800000, 7820000,
       7840000, 7860000, 7880000, 7900000, 7920000, 79400

This provides us with a more detailed view of the data.
For each frame there are 8 different detector angles, such that each frames comprises $50\times50\times8=20,000$ data points.
In the present example, each frame consists of $50\times50$ pixels but in principle the number of pixels can vary from frame to frame.

We furthermore can see that for example for the first frame the principal and secondary rotation angles are 0.0476 radians and 0 radians, respectively.

We can also inspect the dimensions of the frames directly via the `stack.dimensions` property; for example, for the first four frames:

In [7]:
dc.stack.dimensions[:4]

array([[50, 50,  8],
       [50, 50,  8],
       [50, 50,  8],
       [50, 50,  8]])

The `data` field of the stack contains a flattened representation of *all* frames.
The size of this field is thus equal to the sum over all frames of the product of the number of pixels in *i* and *j* and the number of detector angles.
The following cell demonstrates this explicitly.

In [8]:
n = sum(np.prod(k) for k in dc.stack.dimensions)
print('{:32} : {}'.format('Expected number of data points', n))
print('{:32} : {}'.format('Size of the data property', len(dc.stack.data)))

Expected number of data points   : 8340000
Size of the data property        : 8340000


<div class="alert alert-block alert-info">
<b>Tip:</b> In this example, the angles are already given in radians in the input file as can be seen in the data container view above or by printing the <code>dc.angles_in_radians</code> property.
</div>

In [9]:
dc.angles_in_radians

True

<div class="alert alert-block alert-info">
<b>Tip:</b> If the angles are given in degrees in the input file, they can be switched to radians by calling the <code>DataContainer.degrees_to_radians</code> method.
</div>

In [10]:
dc.degrees_to_radians?

Signature:
dc.degrees_to_radians(
    convert_principal: bool = False,
    convert_secondary: bool = False,
    convert_detector: bool = False,
) -> None
Docstring:
Converts angles in data from degrees to radians. Radians are
necessary to run the reconstruction correctly.

Parameters
----------
convert_principal
    If ``True``, converts the principal rotation angles
    from degrees to radians. Default is ``False``.
convert_secondary
    If ``True``, converts the secondary rotation angles
    from degrees to radians. Default is ``False``.
convert_detector
    If ``True``, converts the detector angles
    from degrees to radians. Default is ``False``.
File:      ~/repos/mumott/mumott/data_handling/data_container.py
Type:      method


## Inspecting individual projections

The `stack` property provides a list-like view of the data that allows us to inspect individual frames (i.e., projections) directly.
We can, e.g., check the tenth projection.

In [11]:
dc.stack[10]

InputFrame(data=array([[[0.02, 0.02, 0.03, ..., 0.  , 0.01, 0.02],
        [0.02, 0.01, 0.03, ..., 0.01, 0.04, 0.02],
        [0.06, 0.02, 0.03, ..., 0.03, 0.03, 0.  ],
        ...,
        [0.01, 0.04, 0.01, ..., 0.02, 0.03, 0.02],
        [0.01, 0.03, 0.01, ..., 0.02, 0.01, 0.01],
        [0.02, 0.01, 0.02, ..., 0.02, 0.03, 0.  ]],

       [[0.02, 0.01, 0.04, ..., 0.05, 0.02, 0.01],
        [0.03, 0.06, 0.04, ..., 0.04, 0.03, 0.04],
        [0.01, 0.03, 0.04, ..., 0.  , 0.02, 0.04],
        ...,
        [0.05, 0.01, 0.01, ..., 0.02, 0.05, 0.03],
        [0.01, 0.04, 0.  , ..., 0.01, 0.01, 0.04],
        [0.03, 0.01, 0.02, ..., 0.03, 0.03, 0.02]],

       [[0.05, 0.03, 0.03, ..., 0.01, 0.  , 0.05],
        [0.01, 0.03, 0.02, ..., 0.  , 0.  , 0.02],
        [0.01, 0.05, 0.01, ..., 0.  , 0.02, 0.  ],
        ...,
        [0.05, 0.  , 0.02, ..., 0.03, 0.02, 0.03],
        [0.03, 0.02, 0.01, ..., 0.04, 0.02, 0.01],
        [0.03, 0.03, 0.03, ..., 0.  , 0.03, 0.02]],

       ...,

       [[0.04, 0.04, 0.03, ..., 0.01, 0.02, 0.06],
        [0.02, 0.03, 0.04, ..., 0.05, 0.04, 0.01],
        [0.01, 0.02, 0.01, ..., 0.04, 0.04, 0.05],
        ...,
        [0.02, 0.04, 0.01, ..., 0.03, 0.02, 0.02],
        [0.02, 0.01, 0.01, ..., 0.01, 0.03, 0.02],
        [0.01, 0.  , 0.02, ..., 0.03, 0.  , 0.01]],

       [[0.01, 0.07, 0.  , ..., 0.04, 0.03, 0.03],
        [0.02, 0.03, 0.01, ..., 0.01, 0.02, 0.04],
        [0.05, 0.04, 0.02, ..., 0.02, 0.03, 0.02],
        ...,
        [0.05, 0.02, 0.01, ..., 0.01, 0.05, 0.02],
        [0.02, 0.02, 0.03, ..., 0.02, 0.04, 0.05],
        [0.01, 0.03, 0.04, ..., 0.05, 0.02, 0.02]],

       [[0.01, 0.03, 0.02, ..., 0.03, 0.02, 0.03],
        [0.01, 0.04, 0.01, ..., 0.07, 0.03, 0.02],
        [0.02, 0.04, 0.03, ..., 0.  , 0.04, 0.03],
        ...,
        [0.01, 0.04, 0.03, ..., 0.06, 0.03, 0.03],
        [0.04, 0.04, 0.03, ..., 0.02, 0.03, 0.01],
        [0.01, 0.03, 0.02, ..., 0.05, 0.01, 0.02]]]), diode=array([[1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.],
       ...,
       [1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.]]), weights=array([[1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.],
       ...,
       [1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.]]), principal_rotation=array([0.99959766]), secondary_rotation=array([0.]), j_offset=array([0.]), k_offset=array([0.]))

In [12]:
dc.stack[0]

InputFrame(data=array([[[0.02, 0.  , 0.01, ..., 0.  , 0.03, 0.01],
        [0.02, 0.04, 0.02, ..., 0.01, 0.  , 0.02],
        [0.01, 0.03, 0.02, ..., 0.03, 0.03, 0.04],
        ...,
        [0.03, 0.01, 0.  , ..., 0.03, 0.01, 0.01],
        [0.04, 0.02, 0.02, ..., 0.01, 0.05, 0.01],
        [0.04, 0.03, 0.03, ..., 0.03, 0.05, 0.02]],

       [[0.02, 0.03, 0.03, ..., 0.03, 0.01, 0.03],
        [0.01, 0.01, 0.01, ..., 0.02, 0.04, 0.  ],
        [0.  , 0.01, 0.03, ..., 0.04, 0.03, 0.03],
        ...,
        [0.  , 0.03, 0.  , ..., 0.01, 0.03, 0.01],
        [0.04, 0.02, 0.  , ..., 0.03, 0.01, 0.03],
        [0.03, 0.  , 0.01, ..., 0.02, 0.04, 0.02]],

       [[0.  , 0.04, 0.02, ..., 0.02, 0.02, 0.04],
        [0.  , 0.03, 0.03, ..., 0.  , 0.03, 0.01],
        [0.03, 0.02, 0.02, ..., 0.01, 0.  , 0.02],
        ...,
        [0.02, 0.05, 0.05, ..., 0.01, 0.04, 0.04],
        [0.01, 0.02, 0.04, ..., 0.05, 0.01, 0.02],
        [0.  , 0.04, 0.01, ..., 0.03, 0.03, 0.02]],

       ...,

       [[0.02, 0.01, 0.02, ..., 0.03, 0.02, 0.  ],
        [0.04, 0.01, 0.01, ..., 0.03, 0.07, 0.01],
        [0.04, 0.02, 0.03, ..., 0.03, 0.02, 0.06],
        ...,
        [0.02, 0.03, 0.02, ..., 0.07, 0.04, 0.02],
        [0.  , 0.03, 0.02, ..., 0.01, 0.04, 0.01],
        [0.02, 0.01, 0.01, ..., 0.02, 0.01, 0.02]],

       [[0.01, 0.03, 0.02, ..., 0.02, 0.02, 0.06],
        [0.02, 0.01, 0.04, ..., 0.05, 0.04, 0.05],
        [0.03, 0.02, 0.03, ..., 0.03, 0.01, 0.03],
        ...,
        [0.03, 0.03, 0.01, ..., 0.03, 0.04, 0.04],
        [0.03, 0.05, 0.02, ..., 0.  , 0.02, 0.02],
        [0.01, 0.01, 0.  , ..., 0.02, 0.02, 0.04]],

       [[0.01, 0.04, 0.03, ..., 0.03, 0.03, 0.03],
        [0.01, 0.03, 0.05, ..., 0.04, 0.02, 0.04],
        [0.01, 0.02, 0.02, ..., 0.01, 0.01, 0.03],
        ...,
        [0.02, 0.04, 0.01, ..., 0.01, 0.04, 0.  ],
        [0.05, 0.02, 0.  , ..., 0.03, 0.03, 0.03],
        [0.05, 0.01, 0.  , ..., 0.02, 0.06, 0.03]]]), diode=array([[1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.],
       ...,
       [1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.]]), weights=array([[1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.],
       ...,
       [1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.]]), principal_rotation=array([0.04759989]), secondary_rotation=array([0.]), j_offset=array([0.]), k_offset=array([0.]))

## Transforming the data

It can be useful or necessary to transform the representation of the data.
This is enabled by `TransformParameters` objects, which provides a variety of options.

In [13]:
TransformParameters?

Init signature:
TransformParameters(
    data_sorting: Tuple[int, int, int] = (0, 1, 2),
    data_index_origin: Tuple[int, int] = (0, 0),
    principal_rotation_right_handed: bool = True,
    secondary_rotation_right_handed: bool = True,
    detector_angle_0: Tuple[int, int] = (1, 0),
    detector_angle_right_handed: bool = True,
    offset_positive: Tuple[bool, bool] = (True, True),
) -> None
Docstring:     
Class indicating the coordinate system the data was stored and measured in.
Used to transform it into the coordinate system of the reconstruction.

Parameters
----------
data_sorting : tuple(int, int, int), optional
    Should contain the numbers ``(0, 1, 2)`` (default order). Indicates the order of
    your data indices. ``0`` indicates the position of the index changing orthogonally
    to the principal axis of rotation (``j``). ``1`` indicates the position of the
    index parallel to the principal axis of rotation. ``2`` indicates the position
    of the index where detector i

For example, we can switch the handedness of the principal rotation axes as demonstrated in the next cell.

In [14]:
iframe = 100
frame = dc.stack[iframe]
print(f'Rotation of frame #{iframe} before the transformation:',
      frame.principal_rotation, frame.secondary_rotation)
tp = TransformParameters(principal_rotation_right_handed=False)
dc.transform(tp)
print(f'Rotation of frame #{iframe} after the transformation: ',
      frame.principal_rotation, frame.secondary_rotation)

Rotation of frame #100 before the transformation: [6.01235835] [0.19634954]
Rotation of frame #100 after the transformation:  [-6.01235835] [0.19634954]
